File: Go/hellointerview/01-two-pointers/triangleNumbers.go
package main

import "sort"

func TriangleNumber(nums []int) int {
	sort.Ints(nums)

	count := 0
	for i := len(nums) - 1; i >= 2; i-- {
		left := 0
		right := i - 1

		for left < right {
			if nums[left]+nums[right] > nums[i] {
				count += right - left
				right--
			} else {
				left++
			}
		}
	}

	return count
}

File: Java/hellointerview/01-two-pointers/triangle_numbers.java
import java.util.Arrays;

public class Solution {
  public Integer triangleNumber(int[] heights) {
    Arrays.sort(heights);

    int result = 0;

    for (int i = heights.length - 1; i >= 2; i--) {
      int left = 0;
      int right = i - 1;
      while (left < right) {
        if (heights[left] + heights[right] > heights[i]) {
          count += right - left;
          right--;
        } else {
          left++;
        }
      }
    }
    return count;
  }
}
